File: phases/18-language-paradigms-and-type-theory/14-dsl-design-internal-and-external/code/main.ts
type Expr =
  | { tag: "lit"; n: number }
  | { tag: "add"; a: Expr; b: Expr }
  | { tag: "mul"; a: Expr; b: Expr };

function evalE(e: Expr): number {
  switch (e.tag) {
    case "lit":
      return e.n;
    case "add":
      return evalE(e.a) + evalE(e.b);
    case "mul":
      return evalE(e.a) * evalE(e.b);
  }
}

const program: Expr = { tag: "mul", a: { tag: "add", a: { tag: "lit", n: 2 }, b: { tag: "lit", n: 3 } }, b: { tag: "lit", n: 4 } };
console.log(evalE(program));

File: phases/18-language-paradigms-and-type-theory/14-dsl-design-internal-and-external/code/main.rs
#[derive(Clone, Debug)]
enum Expr {
    Lit(i64),
    Add(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
}

fn eval(e: &Expr) -> i64 {
    match e {
        Expr::Lit(n) => *n,
        Expr::Add(a, b) => eval(a) + eval(b),
        Expr::Mul(a, b) => eval(a) * eval(b),
    }
}

fn main() {
    let p = Expr::Mul(
        Box::new(Expr::Add(Box::new(Expr::Lit(2)), Box::new(Expr::Lit(3)))),
        Box::new(Expr::Lit(4)),
    );
    println!("{}", eval(&p));
}
